Feature j unit testing

pachecowolox opened the pull request:
## Summary
- Add jUnit controller
## Trello Card
https://trello.com/c/J9bQcs4k/28-junit-testing

pachecowolox committed:
add unit test

pachecowolox committed:
add unit test

pachecowolox committed:
add unit test

pachecowolox committed:
add unit test

pachecowolox committed:
add unit test

pachecowolox committed:
add unit test

juanwongramirez commented on src/test/java/wolox/training/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,86 @@
+
+	@Test
+	public void createBooks() throws Exception {
+		Book book = bookBuilder();
```
You can use a class object Book and create an init method with annotation @BeforeEach, the values of the object will be replaced by the values in the init method

juanwongramirez commented on src/test/java/wolox/training/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	@Test
+	public void createBooks() throws Exception {
```
You can use in the name of the tests the convention **given[condition]_when[Action]_then[Expected_reult]**, this must to be a description of the test case, example:

- givenBookWhenAllValidValuesThenStatus200

juanwongramirez commented on src/test/java/wolox/training/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	@Test
+	public void findAll() throws Exception {
```
You can use the test name convention here

juanwongramirez commented on src/test/java/wolox/training/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	@Test
+	public void findByAuthor() throws Exception {
```
You can use the test name convention here

juanwongramirez commented on src/test/java/wolox/training/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	@Test
+	public void findAll() throws Exception {
```
You can use the test name convention here

juanwongramirez commented on src/test/java/wolox/training/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	@Test
+	public void findByUsername() throws Exception {
```
You can use the test name convention here

juanwongramirez commented on src/test/java/wolox/training/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	@Test
+	public void createUser() throws Exception {
```
You can use the test name convention here

juanwongramirez commented:
In the tests you can consider some fail cases or different states of 200

pachecowolox committed:
refactor unit test

pachecowolox committed:
clean code UserRepositoryTest

pachecowolox committed:
refactor unit test UserRepositoryTest

juanwongramirez commented on src/test/java/wolox/training/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,104 @@
+	public void whenCreateBook_thenFindAllUsers() throws Exception {
+		bookBuilder.setAuthor("author1");
+		bookBuilder.setIsbn("isbn1");
+		bookRepository.save(bookBuilder);
```
You can move these tree lines to the init() method and just validate the http responses from the controller

juanwongramirez commented on src/test/java/wolox/training/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	@Test
+	public void whenCreateBook_thenFindByAuthor() throws Exception {
```
You can user more descriptive name, in this methods you need to test the behavior of your controller, an example name is **whenSendValidAuthor_ThenResponseBook**

juanwongramirez commented on src/test/java/wolox/training/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,94 @@
+		user.setName("manuel");
+		user.setUsername("USERNAME1");
+		user.setBirthdate(LocalDate.of(2020, 1, 1));
+		userRepository.save(user);
```
You can use @BeforeEach in this Test to use the save method

marcegl91 commented on src/main/java/wolox/training/models/User.java:
```diff
@@ -36,7 +40,8 @@
 	@Column(nullable = false)
-	@JsonFormat(pattern = "yyyy-MM-dd")
+	@JsonDeserialize(using = LocalDateDeserializer.class)
+	@JsonSerialize(using = LocalDateSerializer.class)
```
Why do you need this?

marcegl91 commented on src/main/resources/application.properties:
```diff
@@ -1,7 +1,10 @@
-spring.datasource.password=123456
-spring.jpa.properties.hibernate.dialect = org.hibernate.dialect.PostgreSQLDialect
+#spring.datasource.url=jdbc:postgresql://localhost:5432/books
+spring.datasource.url = jdbc:h2:mem:books
```
Test should not modify how we run our code. If you keep this code commented, the application will never start. Why do you need to do this?

marcegl91 commented on src/test/java/wolox/training/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,104 @@
+		bookBuilder.setIsbn("isbn1");
+		bookRepository.save(bookBuilder);
+
+		mockMvc.perform(get("/api/books").contentType(MediaType.APPLICATION_JSON)).andExpect(status().isOk())
```
To improve readability, please start each . in a new line, ie:
```
        mvc.perform(get("/api/books")
            .contentType(MediaType.APPLICATION_JSON))
            .etc
```

marcegl91 replied:
Please remember to fix this.

marcegl91 commented on src/test/java/wolox/training/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,104 @@
+
+		bookBuilder.setAuthor("author2");
+		bookBuilder.setIsbn("isbn2");
+		bookRepository.save(bookBuilder);
```
Here you are using another component that you can mock the functionality thus not having a dependency with them (ie, by using Mockito and declaring `private BookRepository bookRepository` with the `@MockBean` annotation).

pachecowolox committed:
refactor unit test book controller and user controller

pachecowolox committed:
change data base application,
change unit test (add mockBean),
delete jackson denpendency

marcegl91 commented on src/main/java/wolox/training/models/User.java:
```diff
@@ -1,6 +1,10 @@
 import com.fasterxml.jackson.annotation.JsonFormat;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.fasterxml.jackson.datatype.jsr310.deser.LocalDateDeserializer;
```
You don't need these imports anymore :) .

marcegl91 commented on build.gradle:
```diff
@@ -23,6 +23,8 @@ dependencies {
 	compile group: 'io.springfox', name: 'springfox-swagger-ui', version: '2.9.2'
 	compile 'org.springframework.plugin:spring-plugin-core:1.2.0.RELEASE'
 	compile group: 'com.google.guava', name: 'guava', version: '27.0-jre'
+	runtimeOnly 'com.h2database:h2'
```
I think you can use testImplementation instead of runtimeOnly: https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_configurations_graph

marcegl91 commented on src/test/java/wolox/training/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,107 @@
+import wolox.training.utils.JsonUtil;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
```
I think you can remove `@SpringBootTest` and `@AutoConfigureMockMvc` and use directly `@WebMvcTest(BookController.class)` instead. Please try it and let us know how it works :) .

marcegl91 commented on src/test/java/wolox/training/controller/UserControllerTest.java:
```diff
@@ -0,0 +1,92 @@
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+@AutoConfigureMockMvc
```
Same as before, with `@WebMvcTest(UserController.class)`

marcegl91 commented on src/test/java/wolox/training/repositories/UserRepositoryTest.java:
```diff
@@ -0,0 +1,54 @@
+		User user = userRepository.findByUsername("PACHECO").orElse(new User());
+
+		//Assert
+		assertThat(user.getUsername().equals(userOneTest.getUsername())).isTrue();
```
You can use `isEqualTo` instead of `equals` and then `isTrue`.

pachecowolox committed:
change assert equals is true to equalsTo,
change runtimeOnly to testImplementation in dependency h2,
clean import model user

pachecowolox committed:
change anotation Spring boot test to @WebMvcTest

marcegl91 commented on src/test/java/wolox/training/repositories/UserRepositoryTest.java:
```diff
@@ -0,0 +1,54 @@
+		//Assert
+		assertThat(user.getUsername()).isEqualTo(userOneTest.getUsername());
+		assertThat(user.getName()).isEqualTo(userOneTest.getName());
+		assertThat(user.getBooks().size() == (userOneTest.getBooks().size())).isTrue();
```
You can use here `isEqualTo`
